src/frontegg-mock/src/lib.rs

```diff
@@ -0,0 +1,280 @@
+#![warn(clippy::suspicious_unary_op_formatting)]
+#![warn(clippy::mut_mutex_lock)]
+#![warn(clippy::print_literal)]
+#![warn(clippy::same_item_push)]
+#![warn(clippy::useless_format)]
+#![warn(clippy::write_literal)]
+#![warn(clippy::redundant_closure)]
+#![warn(clippy::redundant_closure_call)]
+#![warn(clippy::unnecessary_lazy_evaluations)]
+#![warn(clippy::partialeq_ne_impl)]
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::convert::Infallible;
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+use hyper::service::{make_service_fn, service_fn};
+use hyper::{body, Body, Request, Response, Server as HyperServer};
+use jsonwebtoken::EncodingKey;
+use mz_frontegg_auth::{ApiTokenArgs, ApiTokenResponse, Claims, RefreshToken, REFRESH_SUFFIX};
+use mz_ore::now::NowFn;
+use mz_ore::retry::Retry;
+use mz_ore::task::RuntimeExt;
+use tokio::runtime::Runtime;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::JoinHandle;
+use uuid::Uuid;
+
+pub struct FronteggMockServer {
+    pub url: String,
+    pub refreshes: Arc<Mutex<u64>>,
+    pub enable_refresh: Arc<AtomicBool>,
+    pub auth_requests: Arc<Mutex<u64>>,
+    pub role_updates_tx: UnboundedSender<(String, Vec<String>)>,
+    pub handle: JoinHandle<Result<(), hyper::Error>>,
+    _runtime: Arc<Runtime>,
+}
+
+impl FronteggMockServer {
+    // Users is a mapping from (client, secret) -> email address.
+    pub fn start(
+        addr: Option<&SocketAddr>,
+        encoding_key: EncodingKey,
+        tenant_id: Uuid,
+        users: BTreeMap<(String, String), String>,
+        roles: BTreeMap<String, Vec<String>>,
+        now: NowFn,
+        expires_in_secs: i64,
+        latency_secs: Option<u64>,
+    ) -> Result<FronteggMockServer, anyhow::Error> {
+        let (role_updates_tx, role_updates_rx) = unbounded_channel();
+        let refreshes = Arc::new(Mutex::new(0u64));
+        let enable_refresh = Arc::new(AtomicBool::new(true));
+        let auth_requests = Arc::new(Mutex::new(0u64));
+        let context = Context {
+            encoding_key,
+            tenant_id,
+            users,
+            roles,
+            role_updates_rx: Arc::new(Mutex::new(role_updates_rx)),
+            now,
+            expires_in_secs,
+            latency_secs,
+            refresh_tokens: Arc::new(Mutex::new(BTreeMap::new())),
+            refreshes: Arc::clone(&refreshes),
+            enable_refresh: Arc::clone(&enable_refresh),
+            auth_requests: Arc::clone(&auth_requests),
+        };
+
+        let runtime = Arc::new(Runtime::new()?);
```

Comment:
I'm surprised this works! In `main.rs` we're also creating a runtime, e.g. `#[tokio::main]`, and we create a runtime here. I thought this would have panicked with "creating a runtime within a runtime".

If possible I would prefer `fn start(...)` to not spawn its own runtime and instead return the server future, e.g.:
```
pub fn start(...) -> Result<(FronteggMockServer, BoxFuture<'static, ()>), _> { ... }
```
This allows us to be more flexible with how `FronteggMockServer` is used. For example, in `main.rs` we can create our own runtime, but for our integration tests we can spawn the returned future on the test runtime.

Reply:
It might not work, but I recall being able to start the binary and it ran. Yes, doing what you describe would be best. I'd definitely prefer to do that refactor in a followup, as this was copied over from existing code.

Reply:
Sweet, sounds good to me!